File: crates/voidx-tools/src/task_status.rs
//! Task status tool — check child-agent task status.
//!
//! Ported from `src/voidx/tools/task_status.py`.

use crate::base::{Tool, ToolContext, ToolResult};
use crate::error::ToolError;
use async_trait::async_trait;

pub struct TaskStatusTool;

#[async_trait]
impl Tool for TaskStatusTool {
    fn id(&self) -> &'static str {
        "task_status"
    }

    fn description(&self) -> &'static str {
        "Check child-agent task status. Returns status (pending/running/completed/error), current step, elapsed time, and recent output preview. Without task_id, lists all tasks."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "task_id": {
                    "type": "string",
                    "description": "Specific task ID to check. Omit to list all tasks."
                }
            }
        })
    }

    async fn execute(
        &self,
        args: serde_json::Value,
        _ctx: &ToolContext,
    ) -> Result<ToolResult, ToolError> {
        let task_id = args.get("task_id").and_then(|v| v.as_str());

        if let Some(_id) = task_id {
            // In a real implementation, this would check the task state
            Ok(ToolResult::new("No active tasks found.".to_string()))
        } else {
            // List all tasks
            Ok(ToolResult::new("No active tasks.".to_string()))
        }
    }
}

File: src/voidx/tools/task_status.py
"""TaskStatus tool — check worker-persona progress. Claude Code aligned."""

from __future__ import annotations

from pydantic import BaseModel, Field

from voidx.tools.base import BaseTool, model_to_json_schema, ToolContext, ToolResult


class TaskStatusInput(BaseModel):
    task_id: str | None = Field(
        default=None,
        description="Specific task ID to check. If omitted, lists all tasks."
    )


class TaskStatusTool(BaseTool):
    id = "task_status"
    description = (
        "Inspect tasks recorded by the child-agent tracker. With task_id, returns that "
        "task's current tracked status; without task_id, lists all tracked tasks."
    )

    def __init__(self, tracker=None):
        super().__init__()
        self._tracker = tracker

    def parameters_schema(self) -> dict:
        return model_to_json_schema(TaskStatusInput)

    async def execute(self, args: dict, ctx: ToolContext) -> ToolResult:
        try:
            inp = TaskStatusInput.model_validate(args)
        except Exception as exc:
            return ToolResult(output=f"Invalid arguments: {exc}", metadata={"error": True})

        if not self._tracker:
            return ToolResult(
                title="Task: Error",
                output="Task tracker not available.",
                summary="error: tracker unavailable",
                metadata={"error": True, "reason": "no_tracker"},
            )

        if inp.task_id:
            task = self._tracker.get(inp.task_id)
            if not task:
                return ToolResult(
                    title="Task: Error",
                    output=f"Task not found: {inp.task_id}",
                    summary=f"error: task not found: {inp.task_id}",
                    metadata={"error": True, "reason": "not_found", "task_id": inp.task_id},
                )
            return ToolResult(
                title=f"Task {inp.task_id}: {task.status}",
                output=self._tracker.format_status(),
                summary=f"task {inp.task_id}: {task.status}",
                metadata={
                    "task_id": task.id, "agent": task.agent,
                    "status": task.status,
                },
            )

        output = self._tracker.format_status()
        running = self._tracker.list_running()
        return ToolResult(
            title=f"Tasks: {len(running)} running, {len(self._tracker.list_all())} total",
            output=output,
            summary=f"{len(running)} running, {len(self._tracker.list_all())} total",
            metadata={"running": len(running), "total": len(self._tracker.list_all())},
        )
